### PR TITLE
Add siddhi-io-twitter Extension

### PR DESCRIPTION
**Pull request by suganyasuven**
-  To retrieves the real time tweets by giving some filtering options.
- To poll the tweets periodically.

## Automation tests
 - Unit tests 
   > Code coverage - 87%

### CODE REVIEW

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/util/TwitterConstants.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (AnuGayan) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/util/TwitterConstants.java]: Change this in  to 2018

File: component/src/test/resources/testng.xml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
+  ~ Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (AnuGayan) [inline component/src/test/resources/testng.xml]: change this in to 2018

File: README.md [hunk 3]
```diff
@@ -1 +1,132 @@
...
+* Select the app created in step 1 and go to “permission” tab and select “read&write” permission
+* Go to the “keys and access tokens” tab and generate new access token
+* Collect following value from “keys and access tokens tab"
+  * Consumer key
+  * Consumer Secret
+  * Access Token
+  * Access Token Secret
+
+
+**Latest API Docs</a>**
```
Comment (pcnfernando) [inline README.md]: Headings are duplicated

File: README.md [hunk 4]
```diff
@@ -1 +1,132 @@
...
+======================================
+* We are committed to ensuring support for this extension in production. Our unique approach ensures that all support leverages our open development methodology and is provided by the very same engineers who build the technology.
+
+* For more details and to take advantage of this unique opportunity contact us via <a target="_blank" href="http://wso2.com/support?utm_source=gitanalytics&utm_campaign=gitanalytics_Jul17">http://wso2.com/support/</a>.
+
+﻿siddhi-io-twitter
+======================================
+The siddhi-io-twitter extension is an extension to Siddhi which is used to receive event data from Twitter API. It supports json format.
+
+**Prerequisites</a>**
```
Comment (pcnfernando) [inline README.md]: Duplicated content. Please go through the whole content

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterSource.java [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+    }
+
+    /**
+     * Used to validate the parameters.
+     */
+
+    private void validateParameter() {
+        Query.ResultType resultType1 = Query.ResultType.valueOf(this.resultType);
+        if (mode.equalsIgnoreCase(TwitterConstants.MODE_STREAMING)) {
+            for (String s : staticOptionsKeys) {
```
Comment (AnuGayan) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterSource.java]: Shall we rename the s to something more readable like "staticOptionKey", it should be a key right?

File: README.md [hunk 6]
```diff
@@ -1 +1,132 @@
...
+* This extension is shipped by default with WSO2 Stream Processor, if you wish to use an alternative version of this extension you can replace the component <a target="_blank" href="https://github.com/wso2-extensions/siddhi-io-twitter/releases">jar</a> that can be found in the `<STREAM_PROCESSOR_HOME>/lib` directory.
+
+**Using the extension as a <a target="_blank" href="https://wso2.github.io/siddhi/documentation/running-as-a-java-library">java library</a>**
+
+* This extension can be added as a maven dependency along with other Siddhi dependencies to your project.
+
+```
+     <dependency>
+        <groupId>org.wso2.extension.siddhi.io.</groupId>
+        <artifactId>siddhi-io-</artifactId>
```
Comment (Anoukh) [inline README.md]: Please verify and update with the correct group id and artifact id

File: component/pom.xml [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            <artifactId>siddhi-query-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.log4j.wso2</groupId>
+            <artifactId>log4j</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.twitter4j</groupId>
+            <artifactId>twitter4j-stream</artifactId>
+            <version>4.0.6</version>
```
Comment (pcnfernando) [inline component/pom.xml]: Let's define the version as a property

File: README.md [hunk 8]
```diff
@@ -1 +1,132 @@
...
+
+    Architecture List : [architecture@wso2.org](mailto:architecture@wso2.org)
+
+**Support</a>**
+======================================
+* We are committed to ensuring support for this extension in production. Our unique approach ensures that all support leverages our open development methodology and is provided by the very same engineers who build the technology.
+
+* For more details and to take advantage of this unique opportunity contact us via <a target="_blank" href="http://wso2.com/support?utm_source=gitanalytics&utm_campaign=gitanalytics_Jul17">http://wso2.com/support/</a>.
+
+﻿siddhi-io-twitter
```
Comment (Anoukh) [inline README.md]: This section is duplicate?

File: component/pom.xml [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.extension.siddhi.io.twitter.*
+                        </Export-Package>
+                        <Private-Package>
+                            twitter4j.*,
+                            org.wso2.extension.siddhi.io.twitter.util
```
Comment (Anoukh) [inline component/pom.xml]: Verify that this is required or not?

File: component/pom.xml (hunk 7 above)
Comment (pcnfernando) [reply component/pom.xml]: Add the twitter4j dependency in parent and use it here

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterConsumer.java [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+
+/**
+ * This class handles consuming tweets and passing to the stream.
+ * Since enums are inherently serializable and thread-safe, enum Singleton pattern is best way to
+ * create Singleton in Java
+ */
+
```
Comment (pcnfernando) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterConsumer.java]: Remove unnecessary spaces in all places

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterSource.java [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+                        optional = true,
+                        defaultValue = "null",
+                        type = {DataType.STRING}),
+                @Parameter(
+                        name = "polling.interval",
+                        description = "Specifies the period of time (in seconds) to poll tweets periodically",
+                        optional = true,
+                        defaultValue = "3600",
+                        type = {DataType.LONG}),
+        },
```
Comment (pcnfernando) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterSource.java]: Include @ReturnAttribute as well

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterConsumer.java [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+public enum TwitterConsumer {
```
Comment (grainier) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterConsumer.java]: Why is this declared as an Enum?

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterConsumer.java [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+                    log.error("Thread was interrupted during wait : " + e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Listens to the public statuses.
+     */
+    public class TwitterStatusListener implements StatusListener {
```
Comment (Anoukh) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterConsumer.java]: Why is this public? Public class name should be filename

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/util/TwitterConstants.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    public static final String POLLING_SEARCH_LANGUAGE = "language";
+    public static final String POLLING_SEARCH_RESULT_TYPE = "result.type";
+    public static final String POLLING_SEARCH_UNTIL = "until";
+    public static final String POLLING_SEARCH_SINCE = "since";
+    public static final String POLLING_INTERVAL = "polling.interval";
+    private static final String FILTER_LEVEL_NONE = "none";
+    private static final String FILTER_LEVEL_MEDIUM = "medium";
+    private static final String FILTER_LEVEL_LOW = "low";
+    public static final String EMPTY_STRING = "";
+    public static final String COMMA = ",";
```
Comment (pcnfernando) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/util/TwitterConstants.java]: Refactor COMMA as DELIMITER.

File: pom.xml [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+            <dependency>
+                <groupId>org.wso2.extension.siddhi.map.json</groupId>
+                <artifactId>siddhi-map-json</artifactId>
+                <version>${siddhi.map.json.version}</version>
+                <scope>test</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.testng</groupId>
+                <artifactId>testng</artifactId>
+                <version>6.11</version>
```
Comment (Anoukh) [inline pom.xml]: Shall we declare this as a property at the bottom of the pom?

File: README.md [hunk 16]
```diff
@@ -1 +1,132 @@
...
+======================================
+* Go to the https://apps.twitter.com/ and create new App
+* Select the app created in step 1 and go to “permission” tab and select “read&write” permission
+* Go to the “keys and access tokens” tab and generate new access token
+* Collect following value from “keys and access tokens tab"
+  * Consumer key
+  * Consumer Secret
+  * Access Token
+  * Access Token Secret
+
```
Comment (pcnfernando) [inline README.md]: Include a statement saying update the attributes for the extension with these values

File: component/src/test/java/org/wso2/extension/siddhi/io/twitter/source/TestCaseOfTwitterSource.java [hunk 17]
```diff
@@ -0,0 +1,902 @@
...
+                query);
+
+        siddhiAppRuntime.addCallback("query1", new QueryCallback() {
+            @Override
+            public void receive(long timeStamp, Event[] inEvents, Event[] removeEvents) {
+                EventPrinter.print(timeStamp, inEvents, removeEvents);
+                for (Event event : inEvents) {
+                    eventCount.getAndIncrement();
+                    LOG.info(event);
+                    //Assert.assertTrue(event.toString().contains("amazon") || event.toString().contains("Amazon"));
```
Comment (pcnfernando) [inline component/src/test/java/org/wso2/extension/siddhi/io/twitter/source/TestCaseOfTwitterSource.java]: why this line is commented. remove if not needed

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/util/QueryBuilder.java [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+
+public class QueryBuilder {
+    private QueryBuilder() {
+    }
+
+    public static Query createQuery (String q, int count, String language, long sinceId, long maxId, String until,
+                              String since, String resultType, String geoCode, double latitude, double longitude,
+                              double radius, String unitName) {
+        Query query;
+        query = new Query(q);
```
Comment (Anoukh) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/util/QueryBuilder.java]: Can be done in one line?

**Comment by tishan89** [comment]
Please utilize the full PR template. This is a feature PR hence all applicable sections of the PR template should be filed

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterSource.java [hunk 19]
```diff
@@ -0,0 +1,539 @@
...
+                TwitterConstants.EMPTY_STRING);
+        until = optionHolder.validateAndGetStaticValue(TwitterConstants.POLLING_SEARCH_UNTIL,
+                TwitterConstants.EMPTY_STRING);
+        since = optionHolder.validateAndGetStaticValue(TwitterConstants.POLLING_SEARCH_SINCE,
+                TwitterConstants.EMPTY_STRING);
+        resultType = optionHolder.validateAndGetStaticValue(TwitterConstants.POLLING_SEARCH_RESULT_TYPE,
+                "mixed");
+        pollingInterval = Long.parseLong(optionHolder.validateAndGetStaticValue
+                (TwitterConstants.POLLING_INTERVAL, "3600"));
+        query = null;
```
Comment (pcnfernando) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterSource.java]: You can initialize in the top

File: pull_request_template.md [hunk 20]
```diff
@@ -1,52 +1,52 @@
-## Purpose
```
Comment (Anoukh) [inline pull_request_template.md]: Revert any changes done to this file

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterConsumer.java (hunk 12 above)
Comment (suganyasuven) [reply component/src/main/java/org/wso2/extension/siddhi/io/twitter/source/TwitterConsumer.java]: AFAIK, using enum is best way to implement singleton pattern. I refered this one, https://stackoverflow.com/questions/70689/what-is-an-efficient-way-to-implement-a-singleton-pattern-in-java

File: pull_request_template.md (hunk 20 above)
Comment (suganyasuven) [reply pull_request_template.md]: I didn't change anything.

File: component/src/main/java/org/wso2/extension/siddhi/io/twitter/util/Util.java [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * This class contains util methods for the extensions.
+ */
+public class Util {
+    private Util() {
+    }
+
+    private static int length;
+    private static int i;
```
Comment (Anoukh) [inline component/src/main/java/org/wso2/extension/siddhi/io/twitter/util/Util.java]: These two variables should not be class variables IMO

File: component/src/test/java/org/wso2/extension/siddhi/io/twitter/source/TestCaseOfTwitterSource.java [hunk 22]
```diff
@@ -0,0 +1,895 @@
...
+            }
+        });
+
+        siddhiAppRuntime.start();
+        SiddhiTestHelper.waitForEvents(waitTime, 1, eventCount, timeout);
+        Assert.assertTrue(eventArrived.get());
+        siddhiAppRuntime.shutdown();
+    }
+
+    @Test
```
Comment (Anoukh) [inline component/src/test/java/org/wso2/extension/siddhi/io/twitter/source/TestCaseOfTwitterSource.java]: Shall we add the dependsOnMethods attribute to all tests?

File: pom.xml [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+            </dependency>
+            <dependency>
+                <groupId>org.testng</groupId>
+                <artifactId>testng</artifactId>
+                <version>${testng.version}</version>
+                <scope>test</scope>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
+    <build>
```
Comment (Anoukh) [inline pom.xml]: We should add the scm tag to parent pom.
Refer https://github.com/wso2-extensions/siddhi-io-kafka/blob/master/pom.xml#L65

File: pom.xml [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-release-plugin</artifactId>
+                <configuration>
+                    <preparationGoals>clean install</preparationGoals>
```
Comment (Anoukh) [inline pom.xml]: Should be clean install -Pdocumentation-deploy for the doc gen to work.
documentation-deploy profile should be specified in component pom.

Comment (Anoukh) [reply pom.xml]: You can refer https://github.com/wso2-extensions/siddhi-io-kafka/blob/master/component/pom.xml#L98 to specify the documentation-deploy profile

**Pull request merged**